File: worker_largos.py
"""Worker de videos largos.

Este proceso no ejecuta modelos locales. Solo usa FFmpeg en CPU para crear
fragmentos de audio cuando Gemini agota su cuota; la transcripcion siempre la
hace Gemini o Groq.

Pensado para ejecutarse en Google Colab o en un worker persistente.
"""

from __future__ import annotations

import argparse
import io
import json
import os
import shutil
import subprocess
import tempfile
import time
from pathlib import Path
from typing import Any

import google.auth
import requests
from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials
from googleapiclient.discovery import build
from googleapiclient.http import MediaFileUpload, MediaIoBaseDownload, MediaInMemoryUpload


SCOPES = ["https://www.googleapis.com/auth/drive"]
GEMINI_API_KEY = os.getenv("GEMINI_API_KEY", "")
GROQ_API_KEY = os.getenv("GROQ_API_KEY", "")
GEMINI_MODEL = os.getenv("GEMINI_MODEL", "gemini-3.7-flash")
GROQ_MODEL = os.getenv("GROQ_MODEL", "whisper-large-v3-turbo")
LANGUAGE = os.getenv("TRANSCRIPTION_LANGUAGE", "es")
FRAGMENT_SECONDS = int(os.getenv("GROQ_FRAGMENT_SECONDS", "600"))


def load_env_file() -> None:
    """Carga variables simples desde web/.env.local sin sobrescribir el entorno."""
    candidates = [Path(__file__).parent / "web" / ".env.local", Path.cwd() / ".env.local"]
    for path in candidates:
        if not path.exists():
            continue
        for raw in path.read_text(encoding="utf-8").splitlines():
            line = raw.strip()
            if not line or line.startswith("#") or "=" not in line:
                continue
            key, value = line.split("=", 1)
            os.environ.setdefault(key.strip(), value.strip().strip('"').strip("'"))
        break


def require_env(name: str) -> str:
    value = os.getenv(name, "").strip()
    if not value:
        raise RuntimeError(f"Falta la variable {name}")
    return value


def drive_service():
    refresh_token = os.getenv("GOOGLE_REFRESH_TOKEN", "").strip()
    if refresh_token and os.getenv("GOOGLE_CLIENT_ID") and os.getenv("GOOGLE_CLIENT_SECRET"):
        credentials = Credentials(
            token=None,
            refresh_token=refresh_token,
            token_uri="https://oauth2.googleapis.com/token",
            client_id=os.environ["GOOGLE_CLIENT_ID"],
            client_secret=os.environ["GOOGLE_CLIENT_SECRET"],
            scopes=SCOPES,
        )
        credentials.refresh(Request())
    else:
        # En un servidor persistente se usa el refresh token configurado arriba.
        credentials, _ = google.auth.default(scopes=SCOPES)
        if not credentials.valid and credentials.expired and credentials.refresh_token:
            credentials.refresh(Request())
    return build("drive", "v3", credentials=credentials, cache_discovery=False)


def list_job_ids(service, control_folder_id: str) -> list[str]:
    query = f"'{control_folder_id}' in parents and trashed=false and mimeType='application/json'"
    ids: list[str] = []
    page_token = None
    while True:
        response = service.files().list(
            q=query,
            fields="nextPageToken,files(id)",
            pageToken=page_token,
            pageSize=100,
            supportsAllDrives=True,
            includeItemsFromAllDrives=True,
        ).execute()
        ids.extend(item["id"] for item in response.get("files", []))
        page_token = response.get("nextPageToken")
        if not page_token:
            return ids


def ensure_drive_folder(service, name: str, parent_id: str, role: str) -> str:
    query = f"'{parent_id}' in parents and name='{name}' and mimeType='application/vnd.google-apps.folder' and trashed=false"
    response = service.files().list(
        q=query,
        fields="files(id)",
        pageSize=10,
        supportsAllDrives=True,
        includeItemsFromAllDrives=True,
    ).execute()
    if response.get("files"):
        return response["files"][0]["id"]
    folder = service.files().create(
        body={
            "name": name,
            "mimeType": "application/vnd.google-apps.folder",
            "parents": [parent_id],
            "appProperties": {"transcription_role": role},
        },
        fields="id",
        supportsAllDrives=True,
    ).execute()
    return folder["id"]


def read_record(service, record_id: str) -> dict[str, Any]:
    request = service.files().get_media(fileId=record_id, supportsAllDrives=True)
    buffer = io.BytesIO()
    downloader = MediaIoBaseDownload(buffer, request)
    done = False
    while not done:
        _, done = downloader.next_chunk()
    return json.loads(buffer.getvalue().decode("utf-8"))


def update_record(service, record_id: str, record: dict[str, Any], status: str, **extra: Any) -> None:
    record = {**record, "status": status, **extra, "updatedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())}
    payload = json.dumps(record, ensure_ascii=False, indent=2).encode("utf-8")
    properties = {"transcription_job": str(record.get("jobId", "")), "transcription_status": status}
    service.files().update(
        fileId=record_id,
        body={"appProperties": properties},
        media_body=MediaInMemoryUpload(payload, mimetype="application/json", resumable=False),
        supportsAllDrives=True,
        fields="id",
    ).execute()
    record.clear()
    record.update(json.loads(payload.decode("utf-8")))


def download_drive_file(service, file_id: str, destination: Path) -> None:
    request = service.files().get_media(fileId=file_id, supportsAllDrives=True)
    with destination.open("wb") as handle:
        downloader = MediaIoBaseDownload(handle, request, chunksize=8 * 1024 * 1024)
        done = False
        while not done:
            _, done = downloader.next_chunk()


def delete_drive_file(service, file_id: str) -> None:
    try:
        service.files().delete(fileId=file_id, supportsAllDrives=True).execute()
    except Exception as exc:
        print(f"Aviso: no se pudo eliminar Drive {file_id}: {exc}")


def cleanup_temp_drive(service, record: dict[str, Any]) -> None:
    file_id = record.get("driveFileId")
    folder_id = record.get("tempFolderId")
    if folder_id:
        query = f"'{folder_id}' in parents and trashed=false"
        children = service.files().list(
            q=query,
            fields="files(id)",
            pageSize=100,
            supportsAllDrives=True,
            includeItemsFromAllDrives=True,
        ).execute().get("files", [])
        child_ids = {child["id"] for child in children}
        for child_id in child_ids:
            delete_drive_file(service, child_id)
        if file_id and str(file_id) not in child_ids:
            delete_drive_file(service, str(file_id))
        delete_drive_file(service, str(folder_id))
    elif file_id:
        delete_drive_file(service, str(file_id))


def is_quota_error(error: Exception | str) -> bool:
    text = str(error).lower()
    return any(marker in text for marker in ("429", "quota", "resource_exhausted", "rate limit", "too many requests"))


def force_groq_for_test(filename: str) -> bool:
    configured = os.getenv("FORCE_GROQ_FILENAMES", "")
    names = [name.strip().lower() for name in configured.split(",") if name.strip()]
    return any(name in filename.lower() for name in names)


def transcribe_with_gemini(video_path: Path) -> str:
    if not GEMINI_API_KEY:
        raise RuntimeError("Falta GEMINI_API_KEY")
    from google import genai

    client = genai.Client(api_key=GEMINI_API_KEY)
    uploaded = client.files.upload(file=str(video_path))
    try:
        state = getattr(uploaded, "state", None)
        while state and getattr(state, "name", "") in ("PROCESSING", "PROVISIONING"):
            time.sleep(5)
            uploaded = client.files.get(name=uploaded.name)
            state = getattr(uploaded, "state", None)
        if state and getattr(state, "name", "") in ("FAILED", "ERROR"):
            raise RuntimeError(f"Gemini no pudo preparar el video: {state.name}")
        prompt = (
            "Transcribe este video completo en español. Devuelve únicamente la transcripción, "
            "sin resumen ni comentarios. Conserva el orden y separa los cambios de hablante "
            "cuando sean identificables."
        )
        response = client.models.generate_content(model=GEMINI_MODEL, contents=[uploaded, prompt])
        text = (getattr(response, "text", "") or "").strip()
        if not text:
            raise RuntimeError("Gemini devolvió una transcripción vacía")
        return text
    finally:
        try:
            client.files.delete(name=uploaded.name)
        except Exception:
            pass


def transcribe_groq(audio_path: Path) -> str:
    if not GROQ_API_KEY:
        raise RuntimeError("Falta GROQ_API_KEY")
    with audio_path.open("rb") as handle:
        response = requests.post(
            "https://api.groq.com/openai/v1/audio/transcriptions",
            headers={"Authorization": f"Bearer {GROQ_API_KEY}"},
            files={"file": (audio_path.name, handle, "audio/mpeg")},
            data={"model": GROQ_MODEL, "language": LANGUAGE, "response_format": "verbose_json"},
            timeout=900,
        )
    if not response.ok:
        raise RuntimeError(f"Groq HTTP {response.status_code}: {response.text[:500]}")
    return (response.json().get("text") or "").strip()


def transcribe_with_groq(video_path: Path, service, record_id: str, record: dict[str, Any], workdir: Path) -> str:
    audio_pattern = workdir / "fragmento_%05d.mp3"
    subprocess.run(
        ["ffmpeg", "-hide_banner", "-loglevel", "error", "-threads", "1", "-i", str(video_path),
         "-vn", "-map", "0:a:0?", "-ac", "1", "-ar", "16000", "-b:a", "32k", "-f", "segment",
         "-segment_time", str(FRAGMENT_SECONDS), "-reset_timestamps", "1", str(audio_pattern)],
        check=True,
    )
    chunks = sorted(workdir.glob("fragmento_*.mp3"))
    if not chunks:
        raise RuntimeError("FFmpeg no encontró una pista de audio")
    texts: list[str] = []
    update_record(
        service, record_id, record, "processing", provider="groq",
        chunksDone=0, chunksTotal=len(chunks),
        phase=f"Fragmentos listos · {len(chunks)}", progressMode="chunks",
    )
    for index, chunk in enumerate(chunks, 1):
        print(f"  Groq fragmento {index}/{len(chunks)}")
        update_record(
            service, record_id, record, "processing", provider="groq",
            chunksDone=index - 1, chunksTotal=len(chunks),
            phase=f"Transcribiendo fragmento {index} de {len(chunks)}", progressMode="chunks",
        )
        text = transcribe_groq(chunk)
        if text:
            texts.append(text)
        update_record(
            service, record_id, record, "processing", provider="groq",
            chunksDone=index, chunksTotal=len(chunks),
            phase=f"Fragmento {index} de {len(chunks)} listo", progressMode="chunks",
        )
    result = "\n\n".join(texts).strip()
    if result:
        update_record(
            service, record_id, record, "processing", provider="groq",
            chunksDone=len(chunks), chunksTotal=len(chunks),
            phase="Uniendo fragmentos", progressMode="indeterminate",
        )
    if not result:
        raise RuntimeError("Groq devolvió una transcripción vacía")
    return result


def upload_result(service, folder_id: str, filename: str, text: str) -> str:
    payload = io.BytesIO((text.rstrip() + "\n").encode("utf-8"))
    media = MediaInMemoryUpload(payload.getvalue(), mimetype="text/plain", resumable=False)
    result = service.files().create(
        body={"name": filename, "parents": [folder_id], "mimeType": "text/plain"},
        media_body=media,
        fields="id",
        supportsAllDrives=True,
    ).execute()
    return result["id"]


def process_job(service, record_id: str, results_folder_id: str) -> None:
    record = read_record(service, record_id)
    if record.get("status") not in ("queued", "processing") or not record.get("driveFileId"):
        return
    name = str(record.get("filename", "video"))
    print(f"Procesando: {name}")
    update_record(
        service, record_id, record, "processing", provider="gemini",
        chunksDone=0, chunksTotal=0, phase="Descargando desde Drive",
        progressMode="indeterminate", error=None,
    )
    try:
        with tempfile.TemporaryDirectory(prefix="transcriptor-") as temporary:
            workdir = Path(temporary)
            suffix = Path(name).suffix or ".mp4"
            video_path = workdir / f"entrada_video{suffix}"
            download_drive_file(service, str(record["driveFileId"]), video_path)
            try:
                if force_groq_for_test(name):
                    raise RuntimeError("429 quota simulada para prueba")
                update_record(service, record_id, record, "processing", provider="gemini", chunksDone=0, chunksTotal=0, phase="Subiendo video a Gemini", progressMode="indeterminate")
                update_record(service, record_id, record, "processing", provider="gemini", chunksDone=0, chunksTotal=0, phase="Analizando el video con Gemini", progressMode="indeterminate")
                text = transcribe_with_gemini(video_path)
                provider = "gemini"
                print("  Transcripción terminada con Gemini")
            except Exception as gemini_error:
                if not is_quota_error(gemini_error):
                    raise
                print(f"  Gemini agotó la cuota; usando Groq: {gemini_error}")
                text = transcribe_with_groq(video_path, service, record_id, record, workdir)
                provider = "groq"
            result_name = Path(name).stem + ".txt"
            result_folder = str(record.get("resultsFolderId") or results_folder_id)
            update_record(service, record_id, record, "processing", provider=provider, chunksDone=1 if provider == "gemini" else record.get("chunksDone", 0), chunksTotal=1 if provider == "gemini" else record.get("chunksTotal", 0), phase="Guardando TXT", progressMode="indeterminate")
            result_id = upload_result(service, result_folder, result_name, text)
        cleanup_temp_drive(service, record)
        update_record(service, record_id, record, "completed", provider=provider, chunksDone=1, chunksTotal=1, phase="TXT listo", progressMode="complete", resultFileId=result_id, resultName=result_name, error=None)
        print("  Listo; video y carpeta temporal eliminados")
    except Exception as error:
        print(f"  ERROR: {error}")
        try:
            cleanup_temp_drive(service, record)
        finally:
            update_record(service, record_id, record, "error", phase="No se pudo completar", progressMode="complete", error=str(error)[:1000])


def main() -> None:
    load_env_file()
    global GEMINI_API_KEY, GROQ_API_KEY, GEMINI_MODEL, GROQ_MODEL, LANGUAGE, FRAGMENT_SECONDS
    GEMINI_API_KEY = os.getenv("GEMINI_API_KEY", "")
    GROQ_API_KEY = os.getenv("GROQ_API_KEY", "")
    GEMINI_MODEL = os.getenv("GEMINI_MODEL", GEMINI_MODEL)
    GROQ_MODEL = os.getenv("GROQ_MODEL", GROQ_MODEL)
    LANGUAGE = os.getenv("TRANSCRIPTION_LANGUAGE", LANGUAGE)
    FRAGMENT_SECONDS = int(os.getenv("GROQ_FRAGMENT_SECONDS", str(FRAGMENT_SECONDS)))
    parent_id = require_env("GOOGLE_DRIVE_PARENT_FOLDER_ID")
    parser = argparse.ArgumentParser()
    parser.add_argument("--once", action="store_true", help="procesa los trabajos disponibles y termina")
    parser.add_argument("--poll", type=int, default=10, help="segundos entre revisiones")
    args = parser.parse_args()
    service = drive_service()
    control_id = os.getenv("GOOGLE_DRIVE_CONTROL_FOLDER_ID") or ensure_drive_folder(service, "Transcriptor - Control", parent_id, "control")
    results_id = os.getenv("GOOGLE_DRIVE_RESULTS_FOLDER_ID") or ensure_drive_folder(service, "Transcriptor - Resultados", parent_id, "results")
    while True:
        jobs = []
        for record_id in list_job_ids(service, control_id):
            try:
                record = read_record(service, record_id)
                if record.get("status") in ("queued", "processing") and record.get("driveFileId"):
                    jobs.append(record_id)
            except Exception as error:
                print(f"Aviso: no se pudo leer {record_id}: {error}")
        for record_id in jobs:
            process_job(service, record_id, results_id)
        if args.once or not jobs:
            if args.once:
                return
        time.sleep(max(2, args.poll))


if __name__ == "__main__":
    main()

File: worker_largos.mjs
import fs from "node:fs";
import fsp from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import { execFile } from "node:child_process";
import { promisify } from "node:util";

const execFileAsync = promisify(execFile);
const ROOT = path.dirname(new URL(import.meta.url).pathname).replace(/^\/(\w):/, "$1:");
const ENV_FILES = [path.join(ROOT, "web", ".env.local"), path.join(ROOT, ".env.local")];

loadEnv();

const SCOPES = "https://www.googleapis.com/auth/drive";
const DRIVE_API = "https://www.googleapis.com/drive/v3";
const DRIVE_UPLOAD_API = "https://www.googleapis.com/upload/drive/v3/files";
const GEMINI_API = "https://generativelanguage.googleapis.com";
const GEMINI_MODEL = process.env.GEMINI_MODEL || "gemini-3.7-flash";
const GROQ_MODEL = process.env.GROQ_MODEL || "whisper-large-v3-turbo";
const LANGUAGE = process.env.TRANSCRIPTION_LANGUAGE || "es";
const FRAGMENT_SECONDS = Number(process.env.GROQ_FRAGMENT_SECONDS || 600);

function loadEnv() {
  for (const file of ENV_FILES) {
    if (!fs.existsSync(file)) continue;
    for (const raw of fs.readFileSync(file, "utf8").split(/\r?\n/)) {
      const line = raw.trim();
      if (!line || line.startsWith("#") || !line.includes("=")) continue;
      const index = line.indexOf("=");
      const key = line.slice(0, index).trim();
      const value = line.slice(index + 1).trim().replace(/^['"]|['"]$/g, "");
      if (!process.env[key]) process.env[key] = value;
    }
    break;
  }
}

function required(name) {
  const value = String(process.env[name] || "").trim();
  if (!value) throw new Error(`Falta la variable ${name}`);
  return value;
}

function sleep(ms) { return new Promise((resolve) => setTimeout(resolve, ms)); }

async function googleAccessToken() {
  const response = await fetch("https://oauth2.googleapis.com/token", {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      client_id: required("GOOGLE_CLIENT_ID"),
      client_secret: required("GOOGLE_CLIENT_SECRET"),
      refresh_token: required("GOOGLE_REFRESH_TOKEN"),
      grant_type: "refresh_token",
    }),
  });
  const data = await response.json();
  if (!response.ok || !data.access_token) throw new Error("No se pudo renovar el acceso a Google Drive");
  return data.access_token;
}

async function driveRequest(url, options = {}) {
  const token = await googleAccessToken();
  const headers = new Headers(options.headers || {});
  headers.set("Authorization", `Bearer ${token}`);
  return fetch(url.startsWith("http") ? url : `${DRIVE_API}${url}`, { ...options, headers });
}

async function driveJson(url, options = {}) {
  const response = await driveRequest(url, options);
  const text = await response.text();
  let data = {};
  try { data = text ? JSON.parse(text) : {}; } catch { data = { raw: text }; }
  if (!response.ok) throw new Error(data.error?.message || text || `Drive HTTP ${response.status}`);
  return data;
}

async function ensureFolder(name, parentId, role) {
  const query = `'${parentId}' in parents and name='${name.replaceAll("'", "\\'")}' and mimeType='application/vnd.google-apps.folder' and trashed=false`;
  const found = await driveJson(`/files?q=${encodeURIComponent(query)}&fields=files(id,name)&pageSize=10&supportsAllDrives=true&includeItemsFromAllDrives=true`);
  if (found.files?.[0]) return found.files[0];
  return driveJson("/files?supportsAllDrives=true&fields=id,name", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name, mimeType: "application/vnd.google-apps.folder", parents: [parentId], appProperties: { transcription_role: role } }),
  });
}

async function createJsonFile(name, parentId, data, properties) {
  const metadata = await driveJson("/files?supportsAllDrives=true&fields=id", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name, mimeType: "application/json", parents: [parentId], appProperties: properties }),
  });
  await updateJsonFile(metadata.id, data, properties);
  return metadata.id;
}

async function updateJsonFile(fileId, data, properties) {
  const content = JSON.stringify(data, null, 2);
  await driveJson(`${DRIVE_UPLOAD_API}/${fileId}?uploadType=media&supportsAllDrives=true`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: content,
  });
  await driveJson(`/files/${fileId}?supportsAllDrives=true&fields=id`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ appProperties: properties }),
  });
}

async function readRecord(fileId) {
  return driveJson(`/files/${fileId}?alt=media&supportsAllDrives=true`);
}

async function updateRecord(fileId, record, status, extra = {}) {
  const next = { ...record, status, ...extra, updatedAt: new Date().toISOString() };
  await updateJsonFile(fileId, next, { transcription_job: String(next.jobId || ""), transcription_status: status });
  Object.keys(record).forEach((key) => delete record[key]);
  Object.assign(record, next);
}

async function listJobIds(folderId) {
  const query = `'${folderId}' in parents and trashed=false and mimeType='application/json'`;
  const data = await driveJson(`/files?q=${encodeURIComponent(query)}&fields=files(id)&pageSize=100&supportsAllDrives=true&includeItemsFromAllDrives=true`);
  return (data.files || []).map((file) => file.id);
}

async function downloadDriveFile(fileId, destination) {
  const response = await driveRequest(`/files/${fileId}?alt=media&supportsAllDrives=true`);
  if (!response.ok || !response.body) throw new Error(`No se pudo descargar el video desde Drive (${response.status})`);
  const file = fs.createWriteStream(destination);
  await response.body.pipeTo(new WritableStream({ write(chunk) { return new Promise((resolve, reject) => file.write(Buffer.from(chunk), (error) => error ? reject(error) : resolve())); }, close() { file.end(); }, abort(error) { file.destroy(error); } }));
}

async function deleteDriveFile(fileId) {
  const response = await driveRequest(`/files/${encodeURIComponent(fileId)}?supportsAllDrives=true`, { method: "DELETE" });
  if (!response.ok && response.status !== 404) throw new Error(`No se pudo eliminar un archivo temporal (${response.status})`);
}

async function cleanupTemp(record) {
  if (!record.tempFolderId) return;
  const query = `'${record.tempFolderId}' in parents and trashed=false`;
  const data = await driveJson(`/files?q=${encodeURIComponent(query)}&fields=files(id)&pageSize=100&supportsAllDrives=true&includeItemsFromAllDrives=true`);
  for (const child of data.files || []) await deleteDriveFile(child.id);
  await deleteDriveFile(record.tempFolderId);
}

async function uploadResult(folderId, filename, text) {
  const metadata = await driveJson("/files?supportsAllDrives=true&fields=id", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name: filename, parents: [folderId], mimeType: "text/plain" }),
  });
  await driveJson(`${DRIVE_UPLOAD_API}/${metadata.id}?uploadType=media&supportsAllDrives=true`, {
    method: "PATCH",
    headers: { "Content-Type": "text/plain; charset=utf-8" },
    body: `${text.trim()}\n`,
  });
  return metadata.id;
}

function isQuotaError(error) {
  return /429|quota|resource_exhausted|rate limit|too many requests/i.test(String(error));
}

async function geminiTranscribe(videoPath, mimeType, displayName) {
  const apiKey = required("GEMINI_API_KEY");
  const buffer = await fsp.readFile(videoPath);
  const start = await fetch(`${GEMINI_API}/upload/v1beta/files?key=${encodeURIComponent(apiKey)}`, {
    method: "POST",
    headers: {
      "X-Goog-Upload-Protocol": "resumable",
      "X-Goog-Upload-Command": "start",
      "X-Goog-Upload-Header-Content-Length": String(buffer.length),
      "X-Goog-Upload-Header-Content-Type": mimeType,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ file: { display_name: displayName } }),
  });
  if (!start.ok) throw new Error(`Gemini upload start HTTP ${start.status}: ${(await start.text()).slice(0, 500)}`);
  const uploadUrl = start.headers.get("x-goog-upload-url");
  if (!uploadUrl) throw new Error("Gemini no devolvió la URL de subida");
  const finish = await fetch(uploadUrl, {
    method: "POST",
    headers: { "Content-Length": String(buffer.length), "X-Goog-Upload-Offset": "0", "X-Goog-Upload-Command": "upload, finalize" },
    body: buffer,
  });
  const uploaded = await finish.json();
  if (!finish.ok || !uploaded.file?.name || !uploaded.file?.uri) throw new Error(`Gemini upload HTTP ${finish.status}: ${JSON.stringify(uploaded).slice(0, 500)}`);
  const fileName = uploaded.file.name;
  try {
    let file = uploaded.file;
    while (file.state?.name === "PROCESSING" || file.state?.name === "PROVISIONING") {
      await sleep(5000);
      const stateResponse = await fetch(`${GEMINI_API}/v1beta/${fileName}?key=${encodeURIComponent(apiKey)}`);
      file = await stateResponse.json();
    }
    if (["FAILED", "ERROR"].includes(file.state?.name)) throw new Error(`Gemini no pudo preparar el video: ${file.state.name}`);
    const response = await fetch(`${GEMINI_API}/v1beta/models/${encodeURIComponent(GEMINI_MODEL)}:generateContent?key=${encodeURIComponent(apiKey)}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ contents: [{ role: "user", parts: [{ file_data: { mime_type: mimeType, file_uri: uploaded.file.uri } }, { text: "Transcribe este video completo en español. Devuelve únicamente la transcripción, sin resumen ni comentarios. Conserva el orden y separa los cambios de hablante cuando sean identificables." }] }] }),
    });
    const data = await response.json();
    if (!response.ok) throw new Error(`Gemini HTTP ${response.status}: ${JSON.stringify(data).slice(0, 700)}`);
    const text = data.candidates?.[0]?.content?.parts?.map((part) => part.text || "").join("").trim();
    if (!text) throw new Error("Gemini devolvió una transcripción vacía");
    return text;
  } finally {
    await fetch(`${GEMINI_API}/v1beta/${fileName}?key=${encodeURIComponent(apiKey)}`, { method: "DELETE" }).catch(() => undefined);
  }
}

async function groqTranscribe(audioPath) {
  const form = new FormData();
  form.append("file", new Blob([await fsp.readFile(audioPath)], { type: "audio/mpeg" }), path.basename(audioPath));
  form.append("model", GROQ_MODEL);
  form.append("language", LANGUAGE);
  form.append("response_format", "verbose_json");
  const response = await fetch("https://api.groq.com/openai/v1/audio/transcriptions", {
    method: "POST",
    headers: { Authorization: `Bearer ${required("GROQ_API_KEY")}` },
    body: form,
  });
  const data = await response.json();
  if (!response.ok) throw new Error(`Groq HTTP ${response.status}: ${JSON.stringify(data).slice(0, 500)}`);
  return String(data.text || "").trim();
}

async function groqTranscribeVideo(videoPath, recordId, record, workdir) {
  const pattern = path.join(workdir, "fragmento_%05d.mp3");
  await execFileAsync("ffmpeg", ["-hide_banner", "-loglevel", "error", "-threads", "1", "-i", videoPath, "-vn", "-map", "0:a:0?", "-ac", "1", "-ar", "16000", "-b:a", "32k", "-f", "" + "segment", "-segment_time", String(FRAGMENT_SECONDS), "-reset_timestamps", "1", pattern], { maxBuffer: 1024 * 1024 });
  const chunks = (await fsp.readdir(workdir)).filter((name) => /^fragmento_\d+\.mp3$/.test(name)).sort().map((name) => path.join(workdir, name));
  if (!chunks.length) throw new Error("FFmpeg no encontró una pista de audio");
  const texts = [];
  await updateRecord(recordId, record, "processing", { provider: "groq", chunksDone: 0, chunksTotal: chunks.length, phase: `Fragmentos listos · ${chunks.length}`, progressMode: "chunks" });
  for (let index = 0; index < chunks.length; index += 1) {
    const number = index + 1;
    await updateRecord(recordId, record, "processing", { provider: "groq", chunksDone: index, chunksTotal: chunks.length, phase: `Transcribiendo fragmento ${number} de ${chunks.length}`, progressMode: "chunks" });
    const text = await groqTranscribe(chunks[index]);
    if (text) texts.push(text);
    await updateRecord(recordId, record, "processing", { provider: "groq", chunksDone: number, chunksTotal: chunks.length, phase: `Fragmento ${number} de ${chunks.length} listo`, progressMode: "chunks" });
  }
  const result = texts.join("\n\n").trim();
  if (!result) throw new Error("Groq devolvió una transcripción vacía");
  await updateRecord(recordId, record, "processing", { provider: "groq", chunksDone: chunks.length, chunksTotal: chunks.length, phase: "Uniendo fragmentos", progressMode: "indeterminate" });
  return result;
}

async function processJob(recordId, resultsFolderId) {
  const record = await readRecord(recordId);
  if (!["queued", "processing"].includes(record.status) || !record.driveFileId) return;
  const name = String(record.filename || "video");
  console.log(`Procesando: ${name}`);
  await updateRecord(recordId, record, "processing", { provider: "gemini", chunksDone: 0, chunksTotal: 0, phase: "Descargando desde Drive", progressMode: "indeterminate", error: null });
  const workdir = await fsp.mkdtemp(path.join(os.tmpdir(), "transcriptor-"));
  const videoPath = path.join(workdir, `entrada_video${path.extname(name) || ".mp4"}`);
  let provider = "gemini";
  try {
    await downloadDriveFile(record.driveFileId, videoPath);
    try {
      if (String(process.env.FORCE_GROQ_FILENAMES || "").toLowerCase().includes(name.toLowerCase())) throw new Error("429 quota simulada para prueba");
      await updateRecord(recordId, record, "processing", { provider: "gemini", phase: "Subiendo video a Gemini", progressMode: "indeterminate" });
      await updateRecord(recordId, record, "processing", { provider: "gemini", phase: "Analizando el video con Gemini", progressMode: "indeterminate" });
      const mimeType = record.mimeType || "video/mp4";
      var text = await geminiTranscribe(videoPath, mimeType, name);
      console.log("  Transcripción terminada con Gemini");
    } catch (error) {
      if (!isQuotaError(error)) throw error;
      provider = "groq";
      console.log(`  Gemini agotó la cuota; usando Groq: ${error}`);
      text = await groqTranscribeVideo(videoPath, recordId, record, workdir);
    }
    const resultName = `${path.basename(name, path.extname(name))}.txt`;
    await updateRecord(recordId, record, "processing", { provider, phase: "Guardando TXT", progressMode: "indeterminate" });
    const resultId = await uploadResult(String(record.resultsFolderId || resultsFolderId), resultName, text);
    await cleanupTemp(record);
    await updateRecord(recordId, record, "completed", { provider, chunksDone: 1, chunksTotal: 1, phase: "TXT listo", progressMode: "complete", resultFileId: resultId, resultName, error: null });
    console.log("  Listo; video y carpeta temporal eliminados");
  } catch (error) {
    console.error(`  ERROR: ${error}`);
    try { await cleanupTemp(record); } finally { await updateRecord(recordId, record, "error", { phase: "No se pudo completar", progressMode: "complete", error: String(error).slice(0, 1000) }); }
  } finally {
    await fsp.rm(workdir, { recursive: true, force: true });
  }
}

async function main() {
  const parentId = required("GOOGLE_DRIVE_PARENT_FOLDER_ID");
  const control = await ensureFolder("Transcriptor - Control", parentId, "control");
  const results = await ensureFolder("Transcriptor - Resultados", parentId, "results");
  const ids = await listJobIds(control.id);
  const jobs = [];
  for (const id of ids) {
    try {
      const record = await readRecord(id);
      if (["queued", "processing"].includes(record.status) && record.driveFileId) jobs.push(id);
    } catch (error) { console.error(`Aviso: no se pudo leer ${id}: ${error}`); }
  }
  for (const id of jobs) await processJob(id, results.id);
}

const once = process.argv.includes("--once");
if (once) {
  main().catch((error) => { console.error(error); process.exitCode = 1; });
} else {
  const poll = Number(process.env.WORKER_POLL_SECONDS || 10);
  while (true) { await main(); await sleep(Math.max(2, poll) * 1000); }
}
